torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -562,26 +562,22 @@ def moe_infer(self, x, topk_ids, topk_weight):
             dist.all_to_all_single(
                 tokens_per_expert_group, tokens_per_expert, group=self.ep_group
             )
-            torch.sum(
-                tokens_per_expert_group.view(self.ep_size, -1),
-                dim=1,
-                out=self.output_splits,
-            )
-            # print(
-            #     f"EP rank {self.ep_rank} receives {self.output_splits} tokens"
-            #     f"EP rank {self.ep_rank} sends {self.input_splits} tokens"
-            # )
-            # Total received tokens of current rank
-            received = torch.empty(1, dtype=torch.int64, device=x.device)
             # DP to EP token shuffle
             on_device_all_to_all_v(
                 self.token_gather_buf,
-                received,
+                self.output_splits,
                 sorted_tokens,
                 self.input_splits,
                 group=self.ep_group,
             )
+            # Output splits sanity check
+            # expected_splits = torch.sum(
+            #     tokens_per_expert_group.view(self.ep_size, -1),
+            #     dim=1,
+            # )
+            # torch.testing.assert_close(self.output_splits, expected_splits)
             # Received tokens from all other ranks. TODO: use mask instead
+            received = self.output_splits.sum().item()
```

Comment:
this is an h2d sync?

Reply:
It is. I am looking for an mechanism that can consume the `output_splits` directly.
Would your PR be related? https://github.com/pytorch/pytorch/pull/148531
Without that mechanism, I am putting a sync here for now.

Reply:
It's related but it comes with "all splits should be divisible by 16" limitation, for inference maybe this grouped gemm https://github.com/pytorch/FBGEMM/blob/98e9e680c664421a6157d55079e58c7938452798/fbgemm_gpu/experimental/gemm/triton_gemm/grouped_gemm.py#L131 is better.